server/statistics/hot_cache_task.go

```diff
@@ -0,0 +1,190 @@
+	peerInfo   *core.PeerInfo
+	regionInfo *core.RegionInfo
+}
+
+// NewCheckPeerTask creates task to update peerInfo
+func NewCheckPeerTask(peerInfo *core.PeerInfo, regionInfo *core.RegionInfo) FlowItemTask {
+	return &checkPeerTask{
+		peerInfo:   peerInfo,
+		regionInfo: regionInfo,
+	}
+}
+
+func (t *checkPeerTask) taskType() flowItemTaskKind {
+	return checkPeerTaskType
+}
+
+func (t *checkPeerTask) runTask(flow *hotPeerCache) {
+	stat := flow.CheckPeerFlow(t.peerInfo, t.regionInfo)
+	if stat != nil {
+		update(stat, flow)
+	}
+}
+
+type checkExpiredTask struct {
+	region *core.RegionInfo
+}
+
+// NewCheckExpiredItemTask creates task to collect expired items
+func NewCheckExpiredItemTask(region *core.RegionInfo) FlowItemTask {
+	return &checkExpiredTask{
+		region: region,
+	}
+}
+
+func (t *checkExpiredTask) taskType() flowItemTaskKind {
+	return checkExpiredTaskType
+}
+
+func (t *checkExpiredTask) runTask(flow *hotPeerCache) {
+	expiredStats := flow.CollectExpiredItems(t.region)
+	for _, stat := range expiredStats {
+		update(stat, flow)
+	}
+}
+
+type collectUnReportedPeerTask struct {
+	storeID   uint64
+	regionIDs map[uint64]struct{}
+	interval  uint64
+}
+
+// NewCollectUnReportedPeerTask creates task to collect unreported peers
+func NewCollectUnReportedPeerTask(storeID uint64, regionIDs map[uint64]struct{}, interval uint64) FlowItemTask {
+	return &collectUnReportedPeerTask{
+		storeID:   storeID,
+		regionIDs: regionIDs,
+		interval:  interval,
+	}
+}
+
+func (t *collectUnReportedPeerTask) taskType() flowItemTaskKind {
+	return collectUnReportedPeerTaskType
+}
+
+func (t *collectUnReportedPeerTask) runTask(flow *hotPeerCache) {
+	stats := flow.CheckColdPeer(t.storeID, t.regionIDs, t.interval)
+	for _, stat := range stats {
+		update(stat, flow)
+	}
+}
+
+type collectRegionStatsTask struct {
+	minDegree int
+	ret       chan map[uint64][]*HotPeerStat
+}
+
+func newCollectRegionStatsTask(minDegree int) *collectRegionStatsTask {
+	return &collectRegionStatsTask{
+		minDegree: minDegree,
+		ret:       make(chan map[uint64][]*HotPeerStat),
+	}
+}
+
+func (t *collectRegionStatsTask) taskType() flowItemTaskKind {
+	return collectRegionStatsTaskType
+}
+
+func (t *collectRegionStatsTask) runTask(flow *hotPeerCache) {
+	t.ret <- flow.RegionStats(t.minDegree)
+}
+
+type isRegionHotTask struct {
+	region       *core.RegionInfo
+	minHotDegree int
+	ret          chan bool
+}
+
+func newIsRegionHotTask(region *core.RegionInfo, minDegree int) *isRegionHotTask {
+	return &isRegionHotTask{
+		region:       region,
+		minHotDegree: minDegree,
+		ret:          make(chan bool),
+	}
+}
+
+func (t *isRegionHotTask) taskType() flowItemTaskKind {
+	return isRegionHotTaskType
+}
+
+func (t *isRegionHotTask) runTask(flow *hotPeerCache) {
+	t.ret <- flow.isRegionHotWithAnyPeers(t.region, t.minHotDegree)
+}
+
+func (t *isRegionHotTask) waitRet(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return false
+	case r := <-t.ret:
```

Comment:
Does this channel need to close?

Reply:
I don't think it is necessary. Sending an item to a closed channel will cause panic. The channel does not need to be explicitly closed to reclaim the memory.

Reply:
As there are multiple senders for the channel, it is hard to close the channel without mutex in order to prevent data race.